crates/rattler_cache/src/package_cache/mod.rs

```diff
@@ -98,17 +131,151 @@ impl From<Cancelled> for PackageCacheError {
     }
 }
 
+impl From<Cancelled> for PackageCacheLayerError {
+    fn from(_value: Cancelled) -> Self {
+        Self::Cancelled
+    }
+}
+
+impl From<PackageCacheLayerError> for PackageCacheError {
+    fn from(err: PackageCacheLayerError) -> Self {
+        // Convert the PackageCacheLayerError to a LayerError by boxing it
+        PackageCacheError::LayerError(Box::new(err))
+    }
+}
+
+impl PackageCacheLayer {
+    /// Determine if the layer is read-only in the filesystem
+    pub fn is_readonly(&self) -> bool {
+        self.path
+            .metadata()
+            .map(|m| m.permissions().readonly())
+            .unwrap_or(false)
+    }
+
+    /// Validate the package, and throw if invalid.
+    pub async fn validate_or_throw(
+        &self,
+        cache_key: &CacheKey,
+    ) -> Result<CacheLock, PackageCacheLayerError> {
+        let cache_entry = self
+            .packages
+            .get(&cache_key.clone().into())
+            .ok_or(PackageCacheLayerError::PackageNotFound)?
+            .clone();
+        let mut cache_entry = cache_entry.lock().await;
+        let cache_path = self.path.join(cache_key.to_string());
+
+        match validate_package_common::<
+            fn(PathBuf) -> _,
+            Pin<Box<dyn Future<Output = Result<(), _>> + Send>>,
+            std::io::Error,
```

Comment:
Is this needed?

Reply:
Yes unfortunately, as the compiler has to infer the types somehow if we're passing in `None` for `fetch` and `reporter`. 

One way around this would be to assign default types to `F`, `Fut`, `E` during the declaration of `validate_package_common<F, Fut, E>`, but that feature is being phased out of Rust.